File: backend/routers/notes.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
import crud
import schemas
from database import get_db

router = APIRouter(prefix="/api/notes", tags=["notes"])


@router.get("", response_model=List[schemas.NoteResponse])
def list_notes(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return crud.get_notes(db, skip=skip, limit=limit)


@router.post("", response_model=schemas.NoteResponse, status_code=status.HTTP_201_CREATED)
def create_note(note: schemas.NoteCreate, db: Session = Depends(get_db)):
    return crud.create_note(db, note)


@router.get("/{note_id}", response_model=schemas.NoteResponse)
def get_note(note_id: int, db: Session = Depends(get_db)):
    db_note = crud.get_note(db, note_id)
    if db_note is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")
    return db_note


@router.put("/{note_id}", response_model=schemas.NoteResponse)
def update_note(note_id: int, note: schemas.NoteUpdate, db: Session = Depends(get_db)):
    db_note = crud.update_note(db, note_id, note)
    if db_note is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")
    return db_note


@router.patch("/{note_id}", response_model=schemas.NoteResponse)
def patch_note(note_id: int, note: schemas.NotePatch, db: Session = Depends(get_db)):
    db_note = crud.patch_note(db, note_id, note)
    if db_note is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")
    return db_note


@router.delete("/{note_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_note(note_id: int, db: Session = Depends(get_db)):
    deleted = crud.delete_note(db, note_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")

File: frontend/src/api/notes.ts
import axios from "axios";

const api = axios.create({
  baseURL: "",
  headers: { "Content-Type": "application/json" },
});

export interface Note {
  id: number;
  title: string;
  content: string | null;
  created_at: string;
  updated_at: string;
}

export interface NoteCreate {
  title: string;
  content?: string;
}

export interface NoteUpdate {
  title: string;
  content?: string;
}

export interface NotePatch {
  title?: string;
  content?: string;
}

export const getNotes = (): Promise<Note[]> =>
  api.get("/api/notes").then((r) => r.data);

export const getNote = (id: number): Promise<Note> =>
  api.get(`/api/notes/${id}`).then((r) => r.data);

export const createNote = (data: NoteCreate): Promise<Note> =>
  api.post("/api/notes", data).then((r) => r.data);

export const updateNote = (id: number, data: NoteUpdate): Promise<Note> =>
  api.put(`/api/notes/${id}`, data).then((r) => r.data);

export const patchNote = (id: number, data: NotePatch): Promise<Note> =>
  api.patch(`/api/notes/${id}`, data).then((r) => r.data);

export const deleteNote = (id: number): Promise<void> =>
  api.delete(`/api/notes/${id}`).then(() => undefined);
